modules/backendworker/backendworker.go

```diff
@@ -0,0 +1,254 @@
+package backendworker
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/user"
+	backendscheduler_client "github.com/grafana/tempo/modules/backendscheduler/client"
+	"github.com/grafana/tempo/modules/overrides"
+	"github.com/grafana/tempo/modules/storage"
+	"github.com/grafana/tempo/pkg/model"
+	"github.com/grafana/tempo/pkg/tempopb"
+	tempo_util "github.com/grafana/tempo/pkg/util"
+	"github.com/grafana/tempo/pkg/util/log"
+	"github.com/grafana/tempo/tempodb/backend"
+)
+
+const (
+	reasonCompactorDiscardedSpans = "trace_too_large_to_compact"
+)
+
+type BackendWorker struct {
+	services.Service
+
+	cfg              Config
+	store            storage.Store
+	overrides        overrides.Interface
+	backendScheduler tempopb.BackendSchedulerClient
+
+	workerID string
+}
+
+// var tracer = otel.Tracer("modules/backendworker")
+
+func New(cfg Config, schedulerClientCfg backendscheduler_client.Config, store storage.Store, overrides overrides.Interface) (*BackendWorker, error) {
+	s := &BackendWorker{
+		cfg:       cfg,
+		store:     store,
+		overrides: overrides,
+	}
+
+	workerID, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+	s.workerID = workerID
+
+	schedulerClient, err := backendscheduler_client.New(cfg.BackendSchedulerAddr, schedulerClientCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create backend scheduler client: %w", err)
+	}
+	s.backendScheduler = schedulerClient
+
+	s.Service = services.NewBasicService(s.starting, s.running, s.stopping)
+	return s, nil
+}
+
+func (w *BackendWorker) starting(ctx context.Context) error {
+	if w.cfg.Poll {
+		w.store.EnablePolling(ctx, w)
+	}
+
+	return nil
+}
+
+func (w *BackendWorker) running(ctx context.Context) error {
+	// TODO: consider dropping the ticker and rely only on the backoff.
```

Comment:
see the comment below. personally i think you should rely on the backoff only but encapsulate it in a func suggested below. it feels fine to immediately reenter `processCompactionJobs` as soon as it returns but I wouldn't be opposed to ticker either with a very short timeout.

Reply:
Yep, I like relying on just the backoff here.